pkg/client/keystone/keystone.go

```diff
@@ -2,20 +2,94 @@ package keystone
 
 import (
 	"bytes"
+	"crypto/tls"
+	"crypto/x509"
 	"encoding/json"
 	"fmt"
+	"io"
 	"io/ioutil"
 	"net/http"
 
+	contrail "github.com/Juniper/contrail-operator/pkg/apis/contrail/v1alpha1"
+	"github.com/Juniper/contrail-operator/pkg/certificates"
 	"github.com/Juniper/contrail-operator/pkg/client/kubeproxy"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-func NewClient(client *kubeproxy.Client) *Client {
-	return &Client{client: client}
+// NewClient prepares keystone client based on properties of passed keystone CRD.
+func NewClient(kubClient client.Client, scheme *runtime.Scheme, config *rest.Config, k *contrail.Keystone) (*Client, error) {
+	if k.Spec.ServiceConfiguration.ExternalAddress != "" {
```

Comment:
you could split this method to create Client either external, secure, not-secure then you could return a struct I mean sth like that:
```
func NewClient(kubClient client.Client, scheme *runtime.Scheme, config *rest.Config, k *contrail.Keystone) (*Client, error) {
	proxy, err := newProxy(kubClient, scheme, config, k)
	if err != nil {
		return nil, err
	}
	return &Client{
		Client:       proxy,
		KeystoneConf: &k.Spec.ServiceConfiguration,
	}, nil
}

func newProxy(kubClient client.Client, scheme *runtime.Scheme, config *rest.Config, k *contrail.Keystone) (keystoneClient, error) {
	if k.Spec.ServiceConfiguration.ExternalAddress != "" {
		caCertificate := certificates.NewCACertificate(kubClient, scheme, k, k.GetName())
		caBundle, _ := caCertificate.GetCaCert()
		return newExtKeystoneClient(k.Spec.ServiceConfiguration.AuthProtocol, k.Spec.ServiceConfiguration.ExternalAddress, k.Spec.ServiceConfiguration.ListenPort, caBundle), nil
	}
	proxy, err := kubeproxy.New(config)
	if err != nil {
		return nil, fmt.Errorf("failed to create kubeproxy: %v", err)
	}
	if k.Spec.ServiceConfiguration.AuthProtocol == "https" {
		return proxy.NewSecureClientForService(k.Namespace, k.Name+"-service", k.Status.Port), nil
	}
	return proxy.NewClientForService(k.Namespace, k.Name+"-service", k.Status.Port), nil
}```

Reply:
ofc there might be some errors and you can change names, but my suggestion is to split it to make it more readable

Reply:
Thank you, followed your suggestion.